src/adapter/src/coord/appends.rs

```diff
@@ -390,39 +370,51 @@ impl Coordinator {
-            // correctly return control, nor can we simply hang here.
-            // TODO: Clean shutdown.
-            append_fut
-                .await
-                .expect("One-shot dropped while waiting synchronously")
-                .unwrap_or_terminate("cannot fail to apply appends");
-            self.group_commit_apply(timestamp, responses, write_lock_guard, notifies, permit)
-                .await;
-        } else {
-            let internal_cmd_tx = self.internal_cmd_tx.clone();
-            task::spawn(
-                || "group_commit_apply",
-                async move {
-                    if let Ok(response) = append_fut.await {
-                        response.unwrap_or_terminate("cannot fail to apply appends");
+        // Spawn a task to do the table writes.
+        let internal_cmd_tx = self.internal_cmd_tx.clone();
+        let apply_write_fut = self.apply_local_write_shareable(timestamp);
+
+        task::spawn(
+            || "group_commit_apply",
+            async move {
+                // Wait for the writes to complete.
+                append_fut
+                    .await
+                    .expect("One-shot dropped while waiting")
+                    .unwrap_or_terminate("cannot fail to apply appends");
+
+                // Apply the write by marking the timestamp as complete on the timeline.
+                //
+                // Note: `apply_write_fut` uses a "shareable timestamp oracle" under the hood,
+                // which is what allows us to move this future into a separate task. A shareable
+                // oracle isn't always available, which is why we fallback to an internal command.
+                match apply_write_fut {
+                    Some(fut) => fut.await,
+                    None => {
+                        // Trigger a GroupCommitApply, which will run before any user commands since we're
+                        // sending it on the internal command sender.
                         if let Err(e) = internal_cmd_tx.send(Message::GroupCommitApply(
                             timestamp,
                             responses,
                             write_lock_guard,
-                            notifies,
                             permit,
                         )) {
                             warn!("Server closed with non-responded writes, {e}");
                         }
-                    } else {
-                        warn!("Writer terminated with writes in indefinite state");
                     }
                 }
-                .instrument(Span::current()),
-            );
-        }
+
+                // Note: while technically we should have `GroupCommitApply` update the notifies, we
+                // know that internal commands get processed before any user commands, so the writes
+                // are still guaranteed to be observable before any user commands, because we
+                // submitted the GroupCommitApply above.
```

Comment:
I might be misreading, but it seems like the following two comments are trying to say the same thing:

> // Trigger a GroupCommitApply, which will run before any user commands since we're
                        // sending it on the internal command sender.

> // Note: while technically we should have `GroupCommitApply` update the notifies, we
                // know that internal commands get processed before any user commands, so the writes
                // are still guaranteed to be observable before any user commands, because we
                // submitted the GroupCommitApply above.

So it might be a bit easier to maintain if we combined them into a single comment. That's subjective though, so feel free to ignore.

Reply:
Ah ha! Updated the comment and combined the two